src/id-data-controls.tsx

```diff
@@ -0,0 +1,293 @@
+        street_2: '',
+        place: '',
+        country: '',
+        primary: true,
+    },
+};
+
+const IdDataControls = () => {
+    const savedIdData = useMemo(() => new SavedIdData(), []);
+    const [customIdData, setCustomIdData] = useState<IdDataElement[]>([]);
+    const [signature, setSignature] = useState<Signature>({ type: 'text', name: '' });
+    const [fixedIdData, setFixedIdData] = useState(DEFAULT_FIXED_FIELDS);
+
+    const resetSavedIdData = useCallback(() => {
+        savedIdData.getAll().then((id_data) => id_data && setCustomIdData(id_data));
+        savedIdData.getAllFixed().then((fixed_data) => {
+            setFixedIdData(
+                (fixed_data ?? []).reduce(
+                    (acc, elem: IdDataElement) => ({ ...acc, [elem.type]: elem.value }),
+                    DEFAULT_FIXED_FIELDS
+                )
+            );
+        });
+        savedIdData.getSignature().then((signature) => signature && setSignature(signature));
+    }, [savedIdData]);
+
+    useEffect(() => {
+        resetSavedIdData();
+    }, [resetSavedIdData]);
+
+    useEffect(() => {
+        savedIdData.storeArray(fieldsArrayFromFixedData(fixedIdData));
+    }, [savedIdData, fixedIdData]);
+
+    useEffect(() => {
+        savedIdData.storeArray(customIdData, false);
+    }, [savedIdData, customIdData]);
+
+    const handleFixedChange = (type: 'name' | 'birthdate' | 'email' | 'address', value: string | Address) => {
+        setFixedIdData(
+            produce((data) => {
+                if (!isAddress(value)) {
+                    if (type !== 'address') data[type] = value;
+                } else {
+                    if (type === 'address') data[type] = { ...value, primary: true };
+                }
+            })
+        );
+    };
+
+    if (Privacy.isAllowed(PRIVACY_ACTIONS.SAVE_ID_DATA)) {
+        return (
+            <div id="id-data-controls-container" className="narrow-page">
+                <DynamicInputContainer
+                    key="id-data-controls"
+                    id="id-data-controls"
+                    onAddField={(new_field) =>
+                        setCustomIdData(
+                            produce((fields) => {
+                                const last_index =
+                                    customIdData
+                                        .map((f) =>
+                                            parseInt(
+                                                f.desc.match(
+                                                    new RegExp(`^${t('custom-input-desc', 'id-data-controls')} (\\d+)$`)
+                                                )?.[1] ?? '',
+                                                10
+                                            )
+                                        )
+                                        // remove the undefined elements
+                                        .filter((d) => d)
+                                        .sort()
```

Comment:
```suggestion
                                        .sort((a, b) => a - b)
```

Otherwise… :D 

![image](https://user-images.githubusercontent.com/4048809/169660707-c6388fd8-f1f4-4480-bb0a-cd563f64d9f5.png)

Reply:
But other than that, it seems like you've fixed #553. I can't reproduce any of the other issues anymore. \o/

Reply:
Yes, but not quite, because if you reload, fields with the same desc vanish and only the latest edit remains.